pkg/kv/kvserver/kvflowcontrol/replica_rac2/processor.go

```diff
@@ -606,28 +613,38 @@ func (p *processorImpl) makeStateConsistentRaftMuLocked(
 	if termChanged {
 		p.term = state.Term
 	}
+	leadChanged := state.Leader != p.leaderID
+	leaseChanged := state.Leaseholder != p.leaseholderID
 	replicasChanged := p.desc.replicasChanged
+
+	// Detect when we leave or enter leadership. Note that leftLeader and
+	// becameLeader can be true simultaneously, meaning that we left being leader
+	// of the previous term, and became the leader of state.Term.
+	wasLeader := p.isLeader
+	p.isLeader = state.IsLeader
+	leftLeader := wasLeader && (termChanged || !p.isLeader)
+	becameLeader := (!wasLeader || leftLeader) && p.isLeader
+
+	// Check the common case: nothing changed.
+	if !leftLeader && !becameLeader && !leadChanged && !leaseChanged && !replicasChanged && !force {
+		// NB: if termChanged is the only observed change, everything has already
+		// been made consistent (since we updated p.term above), so we will fall
+		// here and return.
+		return
+	}
+	// At least one thing has changed, or the initialization is forced.
+
 	if replicasChanged {
 		p.desc.replicasChanged = false
 	}
-	if !replicasChanged && state.Leader == p.leaderID && state.Leaseholder == p.leaseholderID &&
-		(p.leader.rc == nil || !termChanged) && !force {
```

Comment:
Was there a bug before this change?

The `p.leader.rc == nil` check seems fragile. It depends on the v1/v2 situation, but here it looks like we want to reason about leader in/out changes regardless of v1/v2.

Before we switch to v2, `p.leader.rc == nil` does not mean "we are not the leader".

The more I look at this line, the more I don't understand it. Hopefully the new code makes it dead simple.